drivers/i2c/i2c_gd32.c

```diff
@@ -0,0 +1,684 @@
+		I2C_CTL0(cfg->reg) &= ~I2C_CTL0_ACKEN;
+	}
+
+	I2C_STAT1(cfg->reg);
+
+	if ((data->xfer_len == 1U) &&
+	    (data->current->flags & I2C_MSG_STOP) &&
+	    (data->current->flags & I2C_MSG_READ)) {
+		data->is_stopped = 1U;
+		I2C_CTL0(cfg->reg) |= I2C_CTL0_STOP;
+	}
+
+}
+
+static void i2c_gd32_event_isr(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+	uint32_t stat = I2C_STAT0(cfg->reg);
+
+	if (stat & I2C_STAT0_SBSEND) {
+		if (data->current->flags & I2C_MSG_READ) {
+			I2C_DATA(cfg->reg) = (data->addr1 << 1U) | 1U;
+		} else {
+			I2C_DATA(cfg->reg) = (data->addr1 << 1U) | 0U;
+		}
+	} else if (stat & I2C_STAT0_ADD10SEND) {
+		I2C_DATA(cfg->reg) = data->addr2;
+	} else if (stat & I2C_STAT0_ADDSEND) {
+		i2c_gd32_handle_addsend(dev);
+	} else if (stat & I2C_STAT0_BTC) {
+		i2c_gd32_handle_btc(dev);
+	} else if (stat & I2C_STAT0_RBNE) {
+		i2c_gd32_handle_rbne(dev);
+	} else if (stat & I2C_STAT0_TBE) {
+		i2c_gd32_handle_tbe(dev);
+	}
+}
+
+static void i2c_gd32_error_isr(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+	uint32_t stat = I2C_STAT0(cfg->reg);
+
+	if (stat & I2C_STAT0_BERR) {
+		I2C_STAT0(cfg->reg) &= ~I2C_STAT0_BERR;
+		data->is_berr = 1U;
+	}
+
+	if (stat & I2C_STAT0_LOSTARB) {
+		I2C_STAT0(cfg->reg) &= ~I2C_STAT0_LOSTARB;
+		data->is_lostarb = 1U;
+	}
+
+	if (stat & I2C_STAT0_AERR) {
+		I2C_STAT0(cfg->reg) &= ~I2C_STAT0_AERR;
+		data->is_aerr = 1U;
+	}
+
+	if (data->is_berr ||
+	    data->is_lostarb ||
+	    data->is_aerr) {
+		/* Notify transfer api, error detected. */
+		k_sem_give(&data->sync_sem);
+	}
+}
+
+static void i2c_gd32_xfer_begin(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+
+	k_sem_reset(&data->sync_sem);
+
+	data->is_berr = 0U;
+	data->is_lostarb = 0U;
+	data->is_aerr = 0U;
+	data->is_restart = 0U;
+	data->is_stopped = 0U;
+
+	i2c_gd32_enable(cfg);
+
+	if (data->current->flags & I2C_MSG_READ) {
+		/* Default to enable ACK bit. */
+		I2C_CTL0(cfg->reg) |= I2C_CTL0_ACKEN;
+
+		/* For 2 bytes read, use POAP bit to give NACK for the last data receiving. */
+		if (data->xfer_len == 2U) {
+			I2C_CTL0(cfg->reg) |= I2C_CTL0_POAP;
+		}
+
+		/*
+		 * For read on 10 bits address mode, start condition will happen twice.
+		 * Transfer sequence as below:
+		 *   S addr1+W addr2 S addr1+R
+		 * Use a is_restart flag to cover this case.
+		 */
+		if (data->dev_config & I2C_ADDR_10_BITS) {
+			data->is_restart = 1U;
+			data->current->flags &= ~I2C_MSG_RW_MASK;
+		}
+	}
+
+	I2C_CTL0(cfg->reg) |= I2C_CTL0_START;
+}
+
+static int i2c_gd32_xfer_end(const struct device *dev)
+{
+	struct i2c_gd32_data *data = dev->data;
+	const struct i2c_gd32_config *cfg = dev->config;
+
+	i2c_gd32_disable_interrupts(cfg);
+
+	if (data->is_berr ||
+	    data->is_lostarb ||
+	    data->is_aerr) {
+		goto error;
```

Comment:
This might be one of those cases where instead of the goto doing a little cleanup right here might read nicer.

Consider using bitflags for errors... then you'd perhaps have something like...

```c
if (data->err != 0) {
  i2c_gd32_disable(cfg);
  i2c_gd32_log_err(data->err);
  err = 0U;
  return -EIO;
}
```

Success is then the rest of the function with a single return 0; at the bottom.

No requirement to change this, but it did take me a second look through to catch all the details

Reply:
done.